pkg/catalogv2/git/download.go

```diff
@@ -1,124 +1,125 @@
 package git
 
 import (
-	"crypto/sha256"
-	"encoding/hex"
-	"encoding/pem"
 	"fmt"
-	"net/url"
-	"os"
-	"path/filepath"
-	"regexp"
-	"strings"
 
 	"github.com/rancher/rancher/pkg/settings"
 	corev1 "k8s.io/api/core/v1"
 )
 
-const (
-	stateDir  = "management-state/git-repo"
-	staticDir = "/var/lib/rancher-data/local-catalogs/v2"
-	localDir  = "../rancher-data/local-catalogs/v2" // identical to helm.InternalCatalog
-)
+// Ensure builds the configuration for a should-existing repo and makes sure it is cloned or reseted to the latest commit of given branch
+func Ensure(secret *corev1.Secret, namespace, name, gitURL, branch string, insecureSkipTLS bool, caBundle []byte) error {
+	git, err := gitForRepo(secret, namespace, name, gitURL, insecureSkipTLS, caBundle)
+	if err != nil {
+		return err
+	}
 
-func gitDir(namespace, name, gitURL string) string {
-	staticDir := filepath.Join(staticDir, namespace, name, hash(gitURL))
-	if s, err := os.Stat(staticDir); err == nil && s.IsDir() {
-		return staticDir
+	return git.EnsureClonedRepo(branch)
+}
+
+// EnsureClonedRepo will check if repo is cloned, if not the method will clone and reset to the latest commit.
+// If reseting to the latest commit is not possible it will fetch and try to reset again
+func (er *extendedRepo) EnsureClonedRepo(branch string) error {
+
+	err := er.cloneOrOpen("")
+	if err != nil {
+		return err
 	}
-	localDir := filepath.Join(localDir, namespace, name, hash(gitURL))
-	if s, err := os.Stat(localDir); err == nil && s.IsDir() {
-		return localDir
+
+	// before: g.reset(branch)
+	// Try to reset to the given commit, if success exit
+	localBranchFullName := fmt.Sprintf("refs/heads/%s", branch)
+	err = er.hardReset(localBranchFullName)
+	if err == nil {
+		return nil
 	}
-	return filepath.Join(stateDir, namespace, name, hash(gitURL))
+
+	// before: fetchAndReset(branch)
+	// If we do not have the commit locally, fetch and reset
+	// return er.fetchAndReset(plumbing.ZeroHash, branch)
+	return er.fetchAndReset(branch)
 }
 
+// Head builds the configuration for a new repo which will be cloned for the first time
 func Head(secret *corev1.Secret, namespace, name, gitURL, branch string, insecureSkipTLS bool, caBundle []byte) (string, error) {
```

Comment:
Can we do the same for head and update as well ?

Reply:
I have implemented a fix, please resolve the conversation